README.md
# Kelp

> Developed by pxav (c) 2019-2020

![GitHub release (latest by date including pre-releases)](https://img.shields.io/github/v/release/PXAV/kelp?include_prereleases&label=version&color=green) [![Maven Central](https://maven-badges.herokuapp.com/maven-central/com.github.pxav.kelp/core/badge.svg)](https://maven-badges.herokuapp.com/maven-central/com.github.pxav.kelp/core)  ![GitHub issues](https://img.shields.io/github/issues/PXAV/kelp) ![GitHub pull requests](https://img.shields.io/github/issues-pr/PXAV/kelp) ![GitHub closed pull requests](https://img.shields.io/github/issues-pr-closed/PXAV/kelp) ![Development Indicator](https://img.shields.io/badge/development-active-brightgreen)



### What is Kelp?

Kelp is an open-source framework for Spigot/Bukkit servers to simplify plugin development and increase development speed. This is achieved by reducing boiler-plate code to a minimum, increasing readability to a maximum and avoid version-specific code.



### The problem with the current spigot plugin development process 

The Spigot API has limited or even no backwards compatibility. Material and sound names change almost in every version, NMS packages are named after their version and require reflections to make them compatible, etc. Adjusting old code to new API versions requires lots of time & effort, which is why many developers don't upgrade their plugins anymore, although many server owners still rely on them. So server owners won't upgrade to the latest Minecraft versions and new ideas can't be realized. 

Furthermore, developers have to use lots of workarounds to achieve certain goals. If you want to create a non-flickering sidebar, you have to save the updateable content into a team prefix, put it into  the sidebar and update the prefix. New developers might not understand why there are teams in the scoreboard, which often causes confusion.



### How Kelp solves all these problems 

Kelp is built modularly. It consists of one core module, which only contains Java-logic and no version-specific code, so it will run on every spigot server version. It contains so-called template classes, which are basically dummies with empty methods doing version specific stuff. The server owner then puts a version-implementation module in a specific folder. This file contains implementations of all the template classes from the core module. It is detected by the core module and every template class is replaced with its version implementation class. So the core module can load version implementations for every version and can run on any version without changing any line in the code-base.



In addition it reduces boiler-plate code and increases readability. Here is an example of how easy NPC creation can be with Kelp:

```java
@EventHandler
public void handlePlayerJoin(PlayerJoinEvent event) {
  
  // create new npc instance
  KelpNpc kelpNpc = npcFactory.newKelpNpc()
          .itemInHand(kelpItemFactory.newKelpItem() // the NPC is holding an apple
                  .material(KelpMaterial.APPLE))    // in its hand
          .addTitleLine("§bThis is an epic NPC") // set titles above the npc
          .addTitleLine("§bCreate as many lines as you want")
          .imitateSneaking() // the npc will sneak when the player sneaks
          .followHeadRotation(); // the npc will always look at the player
  
  // finally make the npc visible to the player
  kelpNpc.spawn(event.getPlayer());
    
}
```



Compare this to complicated and version-specific packet handling, reflections and DataWatcher-manipulation:

```java
PlayerConnection playerConnection = ((CraftPlayer)player).getHandle().playerConnection;
WorldServer nmsWorld = ((CraftWorld) player.getWorld()).getHandle();
PacketPlayOutNamedEntitySpawn spawnPacket = new PacketPlayOutNamedEntitySpawn();
int entityId = ThreadLocalRandom.current().nextInt(10_000) + 2_000;

GameProfile gameProfile = new GameProfile(npc.getUuid(), npc.getCustomName());

setValue(spawnPacket, "a", entityId);
setValue(spawnPacket, "b", gameProfile.getId());
setValue(spawnPacket, "c", MathHelper.floor(npc.getLocation().getX() * 32.0D));
setValue(spawnPacket, "d", MathHelper.floor(npc.getLocation().getY() * 32.0D));
setValue(spawnPacket, "e", MathHelper.floor(npc.getLocation().getZ() * 32.0D));
setValue(spawnPacket, "f", (byte) ((int) (npc.getLocation().getYaw() * 256.0F / 360.0F)));
setValue(spawnPacket, "g", (byte) ((int) (npc.getLocation().getPitch() * 256.0F / 360.0F)));
setValue(spawnPacket, "h", npc.getItemInHand().getItemStack().getType().getId());

DataWatcher dataWatcher = new DataWatcher(null);
dataWatcher.a(0, (byte) 0x02);
setValue(spawnPacket, "i", dataWatcher);

playerConnection.sendPacket(spawnPacket);
...

```





## Feature overview

- **Sidebar system:** Create sidebars - no matter if animated or just simple & casual. Use up to 32 chars per line and enjoy flicker-free updating
- **Inventory system:** Create stunning GUIs with minimal effort, use an integrated listener system and meet helpful pre-defined or custom widgets. Never mess with changed material names anymore.
- **NPC system:** Create NPCs with ease and give them abilities like sneaking or give them effects and animations. 
- **Entity system:** Create and modify entities without directly spawning them.
- **Command system:** Create commands and sub-commands with complex structures and easily maintain them.
- **Config system:** Create key-value-based configurations for your messages, etc.
- **KelpConnect:** A simple and easy to use netty wrapper for Kelp  
- **Particle engine:** coming soon



## Support & Requirements

The following requirements are needed:
* Java 8 or higher
* A server with Linux/Windows (MacOS has not been tested sufficiently yet)
* A spigot/paper server with one of the supported versions

There are version implementations for the following version implementations available:

| Server Version | Supported | Notes                                                        |
| -------------- | --------- | ------------------------------------------------------------ |
| 1.0-1.7        | ❌         | At the current state of the project it is not planned to support these versions |
| 1.8            | ✅         | All features included                                        |
| 1.9            | ❌         | Following soon                                               |
| 1.10           | ❌         | Following soon                                               |
| 1.11           | ❌         | Following soon                                               |
| 1.12           | ❌         | Following soon                                               |
| 1.13           | ❌         | Following soon                                               |
| 1.14           | ❌         | Work in progress                                             |
| 1.15           | ❌         | Following soon                                               |

## Downloading

#### Maven
```xml
<dependency>
  <groupId>com.github.pxav.kelp</groupId>
  <artifactId>core</artifactId>
  <version>0.0.3</version>
</dependency>
```

### Gradle
```shell script
implementation 'com.github.pxav.kelp:core:0.0.3'
```

### Bazel
```shell script
maven_jar(
    name = "core",
    artifact = "com.github.pxav.kelp:core:0.0.3",
    sha1 = "4743f29c20f3b033de5fe8c1eddb374511fa31d8",
)
```

### Pre-built files
If you are a server owner who simply needs the jar files or a developer who does not use a built tool like Maven, you can simply download the pre-built jar files from the [Releaes page](https://github.com/PXAV/kelp/releases). There you can find all versions, but it's recommended to use the latest.

### Build from source

To build all Kelp modules you first have to compile all CraftBukkit versions from 1.8 to 1.14 for yourself using the official BuildTools.jar. The CraftBukkit library may not be distributed by Kelp for legal reasons. A tutorial can be found [here](https://www.spigotmc.org/wiki/buildtools/).

```shell
java -jar BuildTools.jar --compile craftbukkit --rev <version>
```



Then, clone the repo from GitHub and build it using maven

```shell
git clone https://github.com/PXAV/kelp.git
cd kelp
mvn install
```



## How to use

Check out the [KelpWiki](https://github.com/PXAV/kelp/wiki), where you can find tutorials about the different features of Kelp. It is not 100% complete but constantly updated. Furthermore most of the classes and methods have JavaDocs, so check them out. A separate JavaDoc page is following soon. 



## Contributing

Currently Kelp is developed by a single developer and bug reports, feature suggestions and pull requests are very welcome. If you want to contribute to the code, have a look at our [contribution guidelines](CONTRIBUTING.md). If you have bug reports, ideas, or need support feel free to open a new issue here on GitHub. 



CHANGELOG.md
# Kelp Changelog

This file contains the changes made in the different releases of the KelpFramework.

The headlines with `MR <number>` stand for the major release. The following headlines mark the minor releases/versions.

## MR 0

#### v0.0.0 (28.09.2020)
* **Beginning of Kelp development**
* First commit to the Git repo

#### v0.0.1 (30.06.2020)
* **First release of Kelp**
* Add entity system
* Add command system
* Add NPC system
* Add configuration system
* Add Netty-Wrapper
* Add sidebar system
* Add entity system
* Add 1.8 support

#### v0.0.2 (03.08.2020)
* **"The inventory update"**
* Add item-bound listener system
* Implement new widgets: `ToggleableWidget`, `ItemWidget`, `Pagination`
* Remove redundant bukkitSubId modifier from `KelpMaterial`
* Interactions (taking out of an inventory, ...) with `KelpItems` are now canceled by default and have to be enabled manually
* Improve/Add documentation for many classes
* Fixed bug that CPU load was on 100% when `AnimatedInventory` was opened.
* Fixed NPE when trying to get item tags of items with type `AIR`

#### v0.0.3 (12.09.2020)
* **The particle engine update**

* Kelp can now be referenced through a build tool like Maven or Gradle as it is now part of the Maven Central Repository.
* Create `ParticleType` class that maps particle type names across different versions.
* Create version template `ParticleVersionTemplate` and a v1.8 implementation for version independent particle spawning
* Create particle effect system to spawn complex particle effects.
* If you request a `ScheduledExecutorService` in a constructor annotated with `@Inject` by Guice, a new scheduled thread pool will be injected automatically.

#### v0.0.4 (date!)

* Add name and description for child modules


src/main/java/de/pxav/kelp/version/KelpVersion.java
package de.pxav.kelp.version;

import java.util.Arrays;

/**
 * A class description goes here.
 *
 * @author pxav
 */
public enum KelpVersion {

  MC_1_8_0(1, "1.8-R0.1-SNAPSHOT"),
  MC_1_8_3(4, "1.8.3-R0.1-SNAPSHOT"),
  MC_1_8_4(5, "1.8.4-R0.1-SNAPSHOT"),
  MC_1_8_5(6, "1.8.5-R0.1-SNAPSHOT"),
  MC_1_8_6(7, "1.8.6-R0.1-SNAPSHOT"),
  MC_1_8_7(8, "1.8.7-R0.1-SNAPSHOT"),
  MC_1_8_8(9, "1.8.8-R0.1-SNAPSHOT"),
  MC_1_9_0(11, "1.9-R0.1-SNAPSHOT"),
  MC_1_9_2(13, "1.9.2-R0.1-SNAPSHOT"),
  MC_1_9_4(15, "1.9.4-R0.1-SNAPSHOT"),
  MC_1_10_0(16, "1.10-R0.1-SNAPSHOT"),
  MC_1_10_2(18, "1.10.2-R0.1-SNAPSHOT"),
  MC_1_11_0(19, "1.11-R0.1-SNAPSHOT"),
  MC_1_11_1(20, "1.11.1-R0.1-SNAPSHOT"),
  MC_1_11_2(21, "1.11.2-R0.1-SNAPSHOT"),
  MC_1_12_0(22, "1.12-R0.1-SNAPSHOT"),
  MC_1_12_1(23, "1.12.1-R0.1-SNAPSHOT"),
  MC_1_12_2(24, "1.12.2-R0.1-SNAPSHOT"),
  MC_1_13_0(25, "1.13-R0.1-SNAPSHOT"),
  MC_1_13_1(26, "1.13.1-R0.1-SNAPSHOT"),
  MC_1_13_2(27, "1.13.2-R0.1-SNAPSHOT"),
  MC_1_14_0(28, "1.14-R0.1-SNAPSHOT"),
  MC_1_14_1(29, "1.14.1-R0.1-SNAPSHOT"),
  MC_1_14_2(30, "1.14.2-R0.1-SNAPSHOT"),
  MC_1_14_3(31, "1.14.3-R0.1-SNAPSHOT"),
  MC_1_14_4(32, "1.14.4-R0.1-SNAPSHOT")
  ;

  private int id;
  private String bukkitVersion;

  KelpVersion(int id, String bukkitVersion) {
    this.id = id;
    this.bukkitVersion = bukkitVersion;
  }

  public int getId() {
    return id;
  }

  public String getBukkitVersion() {
    return bukkitVersion;
  }

  public static KelpVersion higherVersion(KelpVersion original, KelpVersion toCompare) {
    return original.getId() > toCompare.getId() ? original : toCompare;
  }

  public static KelpVersion lowerVersion(KelpVersion original, KelpVersion toCompare) {
    return original.getId() > toCompare.getId() ? toCompare : original;
  }

  public static KelpVersion highestVersion() {
    KelpVersion output = KelpVersion.values()[0];
    for (KelpVersion version : KelpVersion.values()) {
      if (version.getId() > output.getId()) {
        output = version;
      }
    }
    return output;
  }

  public static KelpVersion lowestVersion() {
    KelpVersion output = KelpVersion.values()[0];
    for (KelpVersion version : KelpVersion.values()) {
      if (version.getId() < output.getId()) {
        output = version;
      }
    }
    return output;
  }

  public static KelpVersion nextFullVersion(KelpVersion relatedTo) {
    for (int i = relatedTo.getId(); i < highestVersion().getId() + 1; i++) {
      if (get(i).toString().endsWith("0")) {
        return get(i);
      }
    }
    return lowestVersion();
  }

  public static KelpVersion highestVersionOf(KelpVersion... versions) {
    KelpVersion output = versions[0];
    for (KelpVersion version : versions) {
      if (higherVersion(output, version) == version) {
        output = version;
      }
    }
    return output;
  }

  // null if not supported
  public static KelpVersion withBukkitVersion(String bukkitVersion) {
    for (KelpVersion version : KelpVersion.values()) {
      if (version.getBukkitVersion().equalsIgnoreCase(bukkitVersion)) {
        return version;
      }
    }
    return null;
  }

  public static KelpVersion get(int index) {
    return KelpVersion.values()[index];
  }

}
